wrote DTO for Vacancies and CRUD controllers

ectomorph312 opened the pull request.

ectomorph312 committed:
wrote DTO for Vacancies and CRUD controllers

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+public class VacancyController {
+
+    @Autowired
+    private VacancyService vacancyService;
```
pass dependencies via constructor please

aidar4j replied:
Damir can help with it

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+    private VacancyService vacancyService;
+
+    @GetMapping("/{id}")
+    public ReadVacancyDto getVacancyProfile(@PathVariable Long id) {
```
Profile? Name is getVacancyProfile, but you give ReadVacancyDto.

aidar4j replied:
And more, you can name it like `getById(id)`

ectomorph312 replied:
Changed to getById

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PostMapping("/{id}")
+    public ReadVacancyDto createNewVacancy(@PathVariable Long id,
```
Here is the same, In VacancyController is obvious that you create Vacancies. So you can name method as `create(...)`

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PutMapping("/{id}")
+    public ReadVacancyDto updateVacancy(@PathVariable Long id,
```
same as create

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @GetMapping
+    public Page<ReadVacancyDto> getAllVacancies(@PageableDefault(sort = {"created"}, direction = Sort.Direction.DESC, size = 15)
```
Does Pagination work well?

aidar4j replied:
`getAll(...)` is enough

ectomorph312 replied:
Changed to getAll, pagination improved

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @DeleteMapping("{id}")
+    public void deleteVacancy(@PathVariable Long id) {
```
same naming as create, update, get ...

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/CreateVacancyDto.java:
```diff
@@ -0,0 +1,24 @@
+@Data
+public class CreateVacancyDto {
+
+    @NotNull
```
Does it work?

ectomorph312 replied:
Yeap

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/ReadVacancyDto.java:
```diff
@@ -0,0 +1,16 @@
+    private String description;
+    private String requirements;
+    private Set competencies;
+    private LocalDateTime created;
```
`createdAt` would be better naming

aidar4j replied:
I think it can be also added `updatedAt`

ectomorph312 replied:
Changed

aidar4j commented on src/main/java/io/fortylines/hrcrm/entity/Vacancy.java:
```diff
@@ -1,12 +1,14 @@
 
+@Data
 @Entity
 @Table(name = "vacancy")
```
Table should be named in plural form like `vacancies`

ectomorph312 replied:
Changed

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+    @Autowired
+    private VacancyMapper vacancyMapper;
+
+    @Autowired
```
in constructor

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.stream.Collectors;
+
+@Service
+public class VacancyServiceImpl implements VacancyService {
```
I think there should be added one more layer, VacancyDtoService. You can ask Damir about this layers Controller -> DtoService -> Service -> Repository

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+    @Override
+    public Vacancy getById(Long id) {
+        Optional<Vacancy> optionalVacancy = vacancyRepository.findById(id);
+        return optionalVacancy.orElseThrow(() -> new EntityNotFoundException("User with this id: " + id + " not found"));
```
Change the message of exception

ectomorph312 replied:
Changed

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PostMapping("/{id}")
+    public ReadVacancyDto createNewVacancy(@PathVariable Long id,
```
You don't need id by creating Vacancy here. Id is self generated in DB.

ectomorph312 replied:
here I get the user_id

aidar4j commented on src/main/java/io/fortylines/hrcrm/entity/Competencies.java:
```diff
@@ -3,4 +3,8 @@
 public enum Competencies {
 
     Java, Python, JavaScript, PHP;
```
Enum values should be upper case

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/CreateVacancyDto.java:
```diff
@@ -0,0 +1,24 @@
+    private String requirements;
+
+    @NotNull
+    private String competencies;
```
Maybe here is better to get list of competencies not just string. In string I can write anything I want. But if you require `List<Competency> competencies`, then frontend should send just Compenetcies.

ectomorph312 replied:
Rewrite

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+                .collect(Collectors.toSet());
+
+        for (String competence : competenciesList) {
+            if (competence.equalsIgnoreCase(competencies)) {
```
Wrong logic here. If Frontend sends in competencies "java 8", you won't add this competency because "java" is not equals "java 8". Do you know what I mean? You shouldn't parse any string here, just require enums from frontend.

ectomorph312 replied:
now this check is not needed, the front end should send the competencies

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+            }
+        }
+
+        Vacancy savedVacancy = vacancyRepository.save(vacancy);
```
can be inlined

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+                .collect(Collectors.toSet());
+
+        for (String competencesTemp : competences) {
+            if (competencesTemp.equalsIgnoreCase(newCompetences)) {
```
same as create

ectomorph312 replied:
I didn't understand what to create?

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/Impl/VacancyServiceImpl.java:
```diff
@@ -0,0 +1,146 @@
+
+
+
+
```
blank lineeeeeeeees

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/service/VacancyService.java:
```diff
@@ -0,0 +1,23 @@
+public interface VacancyService {
+
+    ReadVacancyDto getVacancyProfile(Long id);
+
```
no need for blank line between methods

ectomorph312 replied:
+

ectomorph312 replied:
Done

aidar4j commented on src/main/resources/db/migration/V1__Create_DB.sql:
```diff
@@ -20,6 +20,7 @@ CREATE TABLE vacancy (
     vacancy_id int8 GENERATED BY DEFAULT AS IDENTITY,
     completed timestamp,
     created timestamp,
+    modified timestamp,
```
Better to start migrations from zero, but never mind. You have added here modified, but not in entity. Please add them. I prefer names like createdAt, changedAt.

ectomorph312 replied:
Changed

ectomorph312 committed:
added_the_dtoService_layer_and_moved_the_business_logic_there

Changes:
Passing dependencies through the constructor,
Changed the names of the UserController and VacancyController methods,
edited columns in Entity Vacancies (updateAt, createAt), deleted the completed column,
added a layer dtoService and moved the business logic there,
Competencies made by uppercase,
changed the logic in CreateVacancies, now the competencies are stored in List, frontend should only send Competencies,
refactored PageRequest by adding default value of sortBy,
corrected migrations and removed blank lines!

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,50 @@
+        return vacancyDtoService.getById(id);
+    }
+
+    @PostMapping("/{user_id}")
```
You can pass user_id in createVacancyDto

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultUserDtoService.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public ReadUserDto create(CreateUserDto createUserDto) {
+        String username = (createUserDto.getFirstName() + createUserDto.getLastName()).toLowerCase();
```
I think a username should be generated as follow: 3 first letters of first name + 3 first letters of last name + year of registration.

Aidar Aibekov = aidaib2021

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/entity/User.java:
```diff
@@ -37,6 +37,10 @@
     @Enumerated(EnumType.STRING)
     private Set<Role> roles;
```
Let's make it a @ManyToOne relation. One User has just one Role.

ectomorph312 replied:
There are only 3 roles in the system,
1. admin - only 1 user can be with this role
and also the HR role and HeadOfDepartment, why would users with the HR role need the HeadOfDepartment role and vice versa ?

ectomorph312 replied:
Or the hierarchy should be like this:
admin -> headofdepartment -> hr
headofdepartment -> hr
hr.
?

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultUserDtoService.java:
```diff
@@ -0,0 +1,102 @@
+        updateUser.setLastName(updateUserDto.getLastName());
+        updateUser.setPassword(newPassword);
+
+        if (active)
```
meaningless if statement, what do you think?

ectomorph312 replied:
I agree, removed

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultVacancyDtoService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    public ReadVacancyDto create(Long user_id, CreateVacancyDto createVacancyDto) {
```
`user_id `is not java style of naming variables. `userId `

ectomorph312 replied:
done

ectomorph312 replied:
userId is now passed to CreateUserDto

ectomorph312 committed:
Working on comments

In vacancyController, in the create method, deleted the used Id, now it is passed to CreateVacancyDto,
username is now generated as follows: 3 first letters FirstName + 3 first letters Last Name + registration year,
deleted if, in the Update, User entity method

ectomorph312 committed:
Working on comments and added the Role entity

In vacancyController, in the create method, deleted the used Id, now it is passed to CreateVacancyDto,

username is now generated as follows: 3 first letters FirstName + 3 first letters Last Name + registration year,

deleted if, in the Update, User entity method

remade the roles, instead of Enum, now the Role entity, ManyToOne relationships, one user has only one role

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/CreateUserDto.java:
```diff
@@ -17,5 +17,5 @@
 
     @NotNull
-    private String roles;
+    private Long roles;
```
Name it `roleId` please

ectomorph312 replied:
renamed

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/CreateVacancyDto.java:
```diff
@@ -0,0 +1,26 @@
+    @NotNull
+    private List<Competencies> competencies;
+
+    private User userId;
```
Here use Long instead of User

ectomorph312 replied:
done

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/ReadUserDto.java:
```diff
@@ -1,15 +1,14 @@
     private String username;
     private Boolean active;
-    private Set roles;
+    private Role roles;
```
Now just one Role assigned, right? Then call it just `role`

ectomorph312 replied:
changed

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/ReadVacancyDto.java:
```diff
@@ -0,0 +1,17 @@
+    private String requirements;
+    private List<Competencies> competencies;
+    private LocalDateTime createdAt;
+    private String author;
```
Why it is String?

ectomorph312 replied:
in entity Vacancies, there is a String method getAuthorName();

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/UpdateUserDto.java:
```diff
@@ -16,12 +16,8 @@
 
     @NotNull
-    private String roles;
+    private Long roles;
```
Again, `roleId`;

ectomorph312 replied:
+

ectomorph312 replied:
renamed

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultUserDtoService.java:
```diff
@@ -0,0 +1,86 @@
+        createUser.setLastName(createUserDto.getLastName());
+        createUser.setPassword(password);
+        createUser.setUsername(username.replaceAll("\\s+",""));
+        createUser.setRoles(new Role(role));
```
Do not create new Role here, assign existing role from db.

ectomorph312 replied:
done, now there is a separate roleService service

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultUserDtoService.java:
```diff
@@ -0,0 +1,86 @@
+        createUser.setFirstName(createUserDto.getFirstName());
+        createUser.setLastName(createUserDto.getLastName());
+        createUser.setPassword(password);
+        createUser.setUsername(username.replaceAll("\\s+",""));
```
Why it is replaceAll here?

ectomorph312 replied:
deleted

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultUserDtoService.java:
```diff
@@ -0,0 +1,86 @@
+        User updateUser = new User();
+        boolean active = updateUserDto.isActive();
+        Long role = updateUserDto.getRoles();
+        String newPassword = passwordEncoder.encode(updateUserDto.getPassword());
```
I would prefer not updating password in here, please remove it. Update of password is task of separate service.

ectomorph312 replied:
temporarily deleted the password change

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultUserDtoService.java:
```diff
@@ -0,0 +1,86 @@
+        updateUser.setLastName(updateUserDto.getLastName());
+        updateUser.setPassword(newPassword);
+        updateUser.setActive(active);
+        updateUser.setRoles(new Role(role));
```
The same as in create

ectomorph312 replied:
changes

aidar4j commented on src/main/java/io/fortylines/hrcrm/dtoService/Impl/DefaultVacancyDtoService.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    public ReadVacancyDto getById(Long id) {
+        Vacancy responseVacancy = vacancyService.getById(id);
```
I would prefer name responses without word "response" in it. `Vacancy vacancy = vacancyService.getById(id);` Consider it please in other places too.

ectomorph312 replied:
changed

aidar4j commented on src/main/resources/application.properties:
```diff
@@ -1,10 +1,14 @@
-### DB
+server.port=8080
+# DB
 spring.datasource.url=jdbc:postgresql://localhost/hrcrm
```
I guess, it is missing port, isn't it?

ectomorph312 replied:
added

ectomorph312 committed:
Deleting a password change and changes assigning a role

working on comments,
deleted the password change from the User,
Instead of creating a new role, the role is now taken from the database,

ectomorph312 committed:
changed the name

removed the response names

aidar4j commented on src/main/java/io/fortylines/hrcrm/controller/VacancyController.java:
```diff
@@ -0,0 +1,49 @@
+        return vacancyDtoService.create(createVacancyDto);
+    }
+
+    @DeleteMapping("{id}")
```
Missing slash I guess

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/dto/ReadVacancyDto.java:
```diff
@@ -0,0 +1,17 @@
+    private String requirements;
+    private List<Competencies> competencies;
+    private LocalDateTime createdAt;
+    private String  author;
```
Redundant space between String and author

ectomorph312 replied:
deleted

aidar4j commented on src/main/java/io/fortylines/hrcrm/entity/User.java:
```diff
@@ -32,13 +31,12 @@
-    private Set<Role> roles;
+    @ManyToOne
+    @JoinColumn(name = "role_id")
+    private Role roles;
```
Please name it `role` not in plural form

ectomorph312 replied:
Done

aidar4j commented on src/main/java/io/fortylines/hrcrm/entity/User.java:
```diff
@@ -32,13 +31,12 @@
 
-    public boolean isAdmin() {
-        return roles.contains(Role.ADMIN);
+    public boolean isActive() {
```
You can name active as isActive and @Data will generate you getters

ectomorph312 replied:
Changed

aidar4j commented on src/main/resources/db/migration/V1__Create_DB.sql:
```diff
@@ -4,29 +4,29 @@ CREATE TABLE users (
     password varchar(255),
-    username varchar(255),
+    username varchar(255) UNIQUE,
+    role_id int8,
```
Please use BIGINT, not int8. Everywhere where primary and foreign key

ectomorph312 replied:
Done

ectomorph312 committed:
removed the space between the line and the author

aidar4j commented on src/main/resources/db/migration/V1__Create_DB.sql:
```diff
@@ -4,29 +4,29 @@ CREATE TABLE users (
 
-CREATE TABLE candidate (
+CREATE TABLE candidates (
     id int8 GENERATED BY DEFAULT AS IDENTITY,
```
Not generated by default, but use auto_increment

ectomorph312 replied:
changed to BIGSERIAL

aidar4j commented on src/main/resources/db/migration/V1__Create_DB.sql:
```diff
@@ -39,14 +39,14 @@ CREATE TABLE candidate (
-ALTER TABLE IF EXISTS candidate ADD CONSTRAINT
-    vacancy_id_pk FOREIGN KEY (vacancy_id) REFERENCES vacancy;
+ALTER TABLE IF EXISTS candidates ADD CONSTRAINT
+    vacancy_id_pk FOREIGN KEY (vacancy_id) REFERENCES vacancies;
```
You can add constraints when creating table, don't do it separately.

ectomorph312 replied:
changed

ectomorph312 committed:
changed migrations and improvements
